### ISSUE TITLE
Accessing internal hidden state

### ISSUE THREAD

**Issue by rahulk90** [Element-Research/rnn#112]
Hi,

I have a question regarding accessing the internal state of an LSTM module .

Say I have the following:
```lua
require 'rnn'
di,dh,do,T = 10,100,20,3
fastNN = nn.FastLSTM(dh,dh)
local rn = nn.Sequential()
  :add(nn.Linear(di, dh))
  :add(fastNN:clone())
  :add(fastNN:clone())
  :add(nn.Linear(dh, do))
rn = nn.Sequencer(nn.Recursor(rn, T))
--Assume A,B,C correctly defined
O1 = rn:forward({A,B,C}) --F1
O2 = rn:forward({A,B,C}) --F2
```

I'd like to access the final hidden state for the LSTMs for rn after the output O1 is computed. (i.e at the end of F1) and use it to set the initial hidden state of the LSTM just before F2. 

Is there a way to do this?

**Comment by nicholas-leonard** [comment]
Because you aren't calling forget between both calls to `forward`, the initial state of the second forward is already set to the output of the first one. This will demonstrate this : 
```lua
require 'rnn'
di,dh,dout,T = 10,100,20,3
fastNN1 = nn.FastLSTM(dh,dh)
fastNN2 = nn.FastLSTM(dh,dh)
local rn = nn.Sequential()
  :add(nn.Linear(di, dh))
  :add(fastNN1)
  :add(fastNN2)
  :add(nn.Linear(dh, dout))
rn = nn.Sequencer(nn.Recursor(rn, T))
--Assume A,B,C correctly defined
local inputs = {torch.randn(3,di), torch.randn(3,di), torch.randn(3,di)}

O1 = rn:forward(inputs) --F1
O2 = rn:forward(inputs) --F2

print(O2[1]:sum(), O2[2]:sum(), O2[3]:sum())

rn:zeroGradParameters()
rn:forget()

O1 = rn:forward(inputs) --F1
local output1, cell1 = fastNN1.output, fastNN1.cell
local output2, cell2 = fastNN2.output, fastNN2.cell
rn:forget()

fastNN1.prevOutput, fastNN1.prevCell = output1, cell1
fastNN2.prevOutput, fastNN2.prevCell = output2, cell2

O2 = rn:forward(inputs) --F2

print(O2[1]:sum(), O2[2]:sum(), O2[3]:sum())
```

**Comment by rahulk90** [comment]
Hi nicholas, thanks a lot for clarifying that! 

The reason I ask is because I'm trying to learn on a long sequence that doesn't fit on my GPU and I'm hoping to do it in chunks. 

Does the same principle apply for the backward pass?  Said differently, if I have a sequence defined as [inputs,inputs] can I do the following:
```
O1 = rn:forward(inputs) --F1
O2 = rn:forward(inputs) --F2

--compute gradO1,gradO2
rn:backward(inputs,gradO2) 
rn:backward(inputs,gradO1) 
```

Would that persist the gradient with respect to the cells and outputs appropriately?

**Comment by nicholas-leonard** [comment]
@rahulk90 No. Do this instead : 
```lua
-- this will tell the rnn to remember the last output
rn:remember('both')
-- first sequence
O1 = rn:forward(inputs1)
rn:backward(inputs1, grad1)
rn:updateParameters(0.1)
-- second sequence
O1 = rn:forward(inputs1)
rn:backward(inputs1, grad1)
rn:updateParameters(0.1)
-- and so forth and so on
```

**Comment by rahulk90** [comment]
@nicholas-leonard perhaps I'm missing something that happens internally but doesn't the backward pass for the first sequence need gradients with respect to the cell state from the second sequence? In the example above, the first backward pass looks like it wouldn't have access to that. 

Finally, could I use optim with this approach? If so, does this suffice?
```
w,dw = rn:getParameters()
function opfunc(...)
 --forward/backward as above
end
w,dw = optim.rmsprop(opfunc, w)
```
or is there other bookkeeping to do.

**Comment by nicholas-leonard** [comment]
@rahulk90 If you do as you say and forward twice and then backward twice, it is the same as just forwarding one long sequence followed by a backward of the same doubly long sequence. So in terms of your GPU memory constraint, this will use double the memory. This is because a backward pass through a sequence needs to save intermediate input/output states for each time-step.

The only way to save on memory is to split your training into smaller sequences as I demonstrated above. Just make the size of your sequence as large as you can afford to fit on your GPU and call forward -> backward -> update  for each sequence. The remember('both') still has the advantage of using the output of the last sequence as the initial hidden state of your next one.

Yes it works with optim.

**Comment by rahulk90** [comment]
@nicholas-leonard Thanks a lot! I understand what you mean. I'll try this out on some test cases and get back to this.

**Comment by lifengjin** [comment]
Hi, I don't quite follow this. If you define a Sequencer first and do forward, the default value of self.train is nil, so in updateOutput 'self.train ~= false' is going to be true, and because _remember is 'neither' by default, which then calls module:forget()?
